lang/java/avro/src/main/java/org/apache/avro/reflect/ReflectData.java

```diff
@@ -845,20 +859,25 @@ public static Schema makeNullable(Schema schema) {
 
   private static final ConcurrentMap<Class<?>, Field[]> FIELDS_CACHE = new ConcurrentHashMap<>();
 
+  private static final ConcurrentMap<Class<?>, Field[]> NATIVE_FIELDS_CACHE = new ConcurrentHashMap<>();
+
   // Return of this class and its superclasses to serialize.
-  private static Field[] getCachedFields(Class<?> recordClass) {
-    return MapUtil.computeIfAbsent(FIELDS_CACHE, recordClass, rc -> getFields(rc, true));
+  private static Field[] getCachedFields(Class<?> recordClass, boolean orderBy) {
+    return MapUtil.computeIfAbsent(orderBy ? FIELDS_CACHE : NATIVE_FIELDS_CACHE, recordClass,
+        rc -> getFields(rc, true, orderBy));
   }
 
-  private static Field[] getFields(Class<?> recordClass, boolean excludeJava) {
+  private static Field[] getFields(Class<?> recordClass, boolean excludeJava, boolean orderBy) {
```

Comment:
```suggestion
  private static Field[] getFields(Class<?> recordClass, boolean excludeJava, boolean useDeterministicFieldOrder) {
```

As `orderBy` is a boolean, it's not a standard "order by X". Can we please rename it?

More than a name like `sortFields`, the name `useDeterministicFieldOrder` captures the fact that a JVM is free to scramble the field order as it sees fit.

Reply:
addressed.